funflow-checkpoints/src/Control/FunFlow/Checkpoints.hs

```diff
@@ -0,0 +1,188 @@
+import           Data.Function          (flip, ($))
+import           Data.Maybe             (Maybe (..))
+import           Data.Proxy
+import           Data.String            (String)
+import           Data.Tuple             (fst, snd)
+import           Data.Typeable
+import           Text.Show              (Show)
+
+data CheckpointException =
+    -- | A checkpoint has been defined multiple times in the flow
+    RepeatedCheckpoint String
+    -- | A conditional branch has a checkpoint in.
+  | CheckpointInConditional String
+    -- | A branch responsible for catching errors has a checkpoint in.
+  | CheckpointInCatch String
+    -- | No matching checkpoint has been defined in the flow
+  | NoCheckpoint String
+  deriving (Eq, Show)
+
+instance Exception CheckpointException
+
+data CheckpointRef a = CheckpointRef (Proxy a) String
+  deriving Eq
+
+-- | Get the name of a constructed checkpoint
+checkpointName :: CheckpointRef a -> String
+checkpointName (CheckpointRef _ x) = x
+
+-- | A checkpoint arrow transfomer.
+data CheckpointT arr (eff :: * -> * -> *) ex a b where
+  Checkpoint :: Typeable a => CheckpointRef a -> CheckpointT arr eff ex a a
+  Unchecked :: arr eff ex a b -> CheckpointT arr eff ex a b
+  Seq :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex a c
+  Par :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex c d -> CheckpointT arr eff ex (a,c) (b,d)
+  Fanin :: CheckpointT arr eff ex a c -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex (Either a b) c
+  Catch :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex (a,ex) b -> CheckpointT arr eff ex a b
+
+instance Category (arr eff ex) => Category (CheckpointT arr eff ex) where
+  id = Unchecked id
+  (.) = flip Seq
+instance Arrow (arr eff ex) => Arrow (CheckpointT arr eff ex) where
+  arr f = Unchecked $ arr f
+  first f = Par f id
+  second = Par id
+  (***) = Par
+instance ArrowChoice (arr eff ex) => ArrowChoice (CheckpointT arr eff ex ) where
+  f +++ g = (f >>> arr Left) ||| (g >>> arr Right)
+  f ||| g = Fanin f g
+instance ArrowError ex (arr eff ex) => ArrowError ex (CheckpointT arr eff ex) where
+  f `catch` g = Catch f g
+
+instance ArrowFlow eff ex (arr eff ex) => ArrowFlow eff ex (CheckpointT arr eff ex) where
+  step' props = Unchecked . step' props
+  stepIO' props = Unchecked . stepIO' props
+  external = Unchecked . external
+  wrap' p eff = Unchecked $ wrap' p eff
+  putInStore = Unchecked . putInStore
+  getFromStore = Unchecked . getFromStore
+  internalManipulateStore = Unchecked . internalManipulateStore
+
+-- | Construct a checkpoint in a flow.
+checkpoint :: Typeable a => String -> CheckpointT arr eff ex a a
+checkpoint = Checkpoint . CheckpointRef (Proxy :: Proxy a)
+
+-- | Run a checkpointed flow to completion
+runToCompletion :: (Arrow (arr eff ex), ArrowChoice (arr eff ex), ArrowError ex (arr eff ex))
+                => CheckpointT arr eff ex a b -> arr eff ex a b
```

Comment:
On the other hand, you need the constraints here again.

Though: should the constraint be in the type constructors? (like `Fanin` stores a proof that the underlying arrow is `ArrowChoice`?). So that we can always `runToCompletion`, but not always use all the constructors?

Reply:
That's a nice idea.